From akpm@osdl.org Thu Aug 31 22:02:16 2006
Message-Id: <200609010502.k8152BF5006264@shell0.pdx.osdl.net>
Subject: PM: device_suspend/resume may sleep
To: mm-commits@vger.kernel.org
Cc: pavel@suse.cz, greg@kroah.com
From: Pavel Machek <pavel@suse.cz>
Date: Thu, 31 Aug 2006 22:02:11 -0700

From: Pavel Machek <pavel@suse.cz>

This adds warning when someone tries them from atomic context.

Signed-off-by: Pavel Machek <pavel@suse.cz>
Signed-off-by: Andrew Morton <akpm@osdl.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/base/power/resume.c  |    1 +
 drivers/base/power/suspend.c |    1 +
 2 files changed, 2 insertions(+)

--- gregkh-2.6.orig/drivers/base/power/resume.c
+++ gregkh-2.6/drivers/base/power/resume.c
@@ -96,6 +96,7 @@ void dpm_resume(void)
 
 void device_resume(void)
 {
+	might_sleep();
 	down(&dpm_sem);
 	dpm_resume();
 	up(&dpm_sem);
--- gregkh-2.6.orig/drivers/base/power/suspend.c
+++ gregkh-2.6/drivers/base/power/suspend.c
@@ -140,6 +140,7 @@ int device_suspend(pm_message_t state)
 {
 	int error = 0;
 
+	might_sleep();
 	down(&dpm_sem);
 	down(&dpm_list_sem);
 	while (!list_empty(&dpm_active) && error == 0) {
